src/c-tips/c-tips.md

```diff
@@ -1,3 +1,209 @@
+give you further control over this process, although care should be taken here
+as well — the compiler will automatically inline functions from the same crate
+where appropriate, so forcing it to do so inappropriately might actually lead
+to decreased performance.
+
+[attributes]: https://doc.rust-lang.org/reference/attributes.html#inline-attribute
+
+Explaining the entire Rust macro system is out of scope for this tips page, so
+you are encouraged to consult the Rust documentation for full details.
+
+## Build System
+
+Most Rust crates are built using Cargo (although it is not required). This
+takes care of many difficult problems with traditional build systems. However,
+you may wish to customise the build process. Cargo provides [`build.rs`
+scripts] for this purpose. They are Rust scripts which can interact with the
+Cargo build system as required.
+
+[`build.rs` scripts]: https://doc.rust-lang.org/cargo/reference/build-scripts.html
+
+Common use cases for build scripts include:
+
+* provide build-time information, for example statically embedding the build
+  date or Git commit hash into your executable
+* generate linker scripts at build time depending on selected features or other
+  logic
+* change the Cargo build configuration
+* add extra static libraries to link against
+
+### Cross-Compiling
+
+Using Cargo for your build system also simplifies cross-compiling. In most
+cases it suffices to tell Cargo `--target thumbv6m-none-eabi` and find a
+suitable executable in `target/thumbv6m-none-eabi/debug/myapp`.
+
+For platforms not natively supported by Rust, you will need to build `libcore`
+for that target yourself. On such platforms, [Xargo] can be used as a stand-in
+for Cargo which automatically builds `libcore` for you.
+
+[Xargo]: https://github.com/japaric/xargo
+
+## Iterators vs Array Access
+
+In C you are probably used to accessing arrays directly by their index:
+
+```c
+int16_t arr[16];
+int i;
+for(i=0; i<sizeof(arr)/sizeof(arr[0]); i++) {
+    process(arr[i]);
+}
+```
+
+In Rust this is an anti-pattern: indexed access can be slower (as it needs to
+be bounds checked) and may prevent various compiler optimisations.
+
+Instead, use iterators:
+
+```rust
+let arr = [0u16; 16];
+for element in arr.iter() {
+    process(*arr);
+}
+```
+
+Iterators provide a powerful array of functionality you would have to implement
+manually in C, such as chaining, zipping, enumerating, finding the min or max,
+summing, and more. Iterator methods can also be chained, giving very readable
+data processing code.
+
+See the [Iterators in the Book] and [Iterator documentation] for more details.
+
+[Iterators in the Book]: https://doc.rust-lang.org/book/second-edition/ch13-02-iterators.html
+[Iterator method documentation]: https://doc.rust-lang.org/core/iter/trait.Iterator.html
+
+## References vs Pointers
+
+In Rust, pointers (called [_raw pointers_]) exist but are rarely used:
```

Comment:
I wouldn't deem this statement accurate. Especially with embedded Rust raw pointer have to be used in many low level places like memory mapped registers and DMA.

Reply:
Good point, I've updated to indicate this.